core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -178,6 +178,7 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Sym
         private final TypeProvider types;
         private final ExpressionEquivalence expressionEquivalence;
         private final boolean dynamicFiltering;
+        private static final int MAX_EXPR_REFERENCE_COUNT = 50;
```

Comment:
What's the reasoning behind 50?
Why not 10? Why not 100?

Reply:
Thanks @findepi for the review. I have to admit there is no science behind that number :) It's based on common query patterns we observed. Normally queries won't have more than 50 expression references, and if they do, I think we should stop there, which is consistent to the original intention of https://github.com/prestodb/presto/pull/10860

Reply:
Would eg 5 be enough? (before the change we had limit of 1, right?)

(just askjing, i still don't fully appreciate the change)

Reply:
I would say 5 is fine with 90% of the cases (some queries just have more filters than others). Yes you're right, without this change the limit is equivalent to 1.

Reply:
IIRC, the limit of 1 was not incidental. I don't remember though, whether it was driven by a concern drawn from real life, or baed on technically possible pathological case. i wish we have had documented that.

@martint @sopel39 do you remember?

Reply:
The limit was set to 1 because anything larger than that could cause exponential expansion in some cases. This was the original issue: https://github.com/prestodb/presto/issues/10455

Reply:
Martin is right. The purpose of the PR is to relax that restriction, in a sense we still avoid exponential expansion, but allow linear expansion (much safer), so that there can be multiple references to the same complex expression which is a common use case

Reply:
I hope the above context helps. Could any of you take another look? Thanks.

Reply:
What's the impact of this change on the query listed in prestodb/presto#10455 ?

Reply:
Sorry for the delay @martint. I'm not sure if other optimization kicks in, but the query from the above ticket produces a plan like this. Maybe this special case is no longer a problem?
```
Output[v]
│   Layout: [expr_13:integer]
│   v := expr_13
└─ Filter[filterPredicate = ("expr_13" = 0)]
   │   Layout: [expr_13:integer]
   └─ Values
          Layout: [expr_13:integer]
          (1)
```

Reply:
Please let me know your thoughts on this @martint Thanks.

Reply:
Apologies for the delay. I'm getting back to this PR now.

For the query above, yes, there are constant folding optimizations in recent version that get rid of all those intermediates. You can adjust the query to:

```sql
WITH
t1 (v) AS (SELECT orderkey FROM orders),
t2 AS( select if(v = 0, v, v) v from t1 ),
t3 AS( select if(v = 0, v, v) v from t2 ),
t4 AS( select if(v = 0, v, v) v from t3 ),
t5 AS( select if(v = 0, v, v) v from t4 ),
t6 AS( select if(v = 0, v, v) v from t5 ),
t7 AS( select if(v = 0, v, v) v from t6 ),
t8 AS( select if(v = 0, v, v) v from t7 )
select *
from t8
where v = 0;
```

However, I realized why it doesn't have the exponential behavior I was worried about. It's because you're checking for `SymbolsExtractor.extractAll(expr).size() <= 1`. It would have been a different story if it had been `extractUnique`, instead.

The downside I see with this PR is a case like the following:

```java
EXPLAIN WITH
    t0 AS (SELECT expensive_function(c) x FROM t),
SELECT *
FROM t0
WHERE x = 1 OR x = 2 OR x = 3 OR x = 4 OR x = 5 OR x = 6 ....
```

In that scenario, the expensive function would be evaluated up to 50 times per row, instead of just once. (This trivial case could clearly be optimized to `x IN (1, 2, 3, ...)`, though).

Also, if I'm reading the code correctly, it wouldn't work if the call to were `expensive_function(c, c)`.